### PR TITLE
Implemented PlatformXchg and add_and_fetch

### PR DESCRIPTION
**Pull request by DmKazakov**

### CODE REVIEW

File: src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp
Comment (AntonKozlov) [inline src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp]: This thing is critical.

Here, `t0` is changed, but the inline assembler block doesn't declare such behavior. This may lead to C code around placing a value to `t0`, that will be destroyed in the asm block, and then following C code assume the value is unchanged, leading to disaster.

Syntax for the block described here: https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html
You can add t0 to "clobbered" list or, better, allow compiler to select a temp register for you.

By example, `add_value` is stored somehow in C code. It's provided in "input" list and constraint that it requested to be in some register by `"r"` specifier. Compiler selects optimal register (most likely -- free at this point)

File: src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp
Comment (AntonKozlov) [inline src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp]: To discsuss.

Can't we use AMOxxx anyway?
```
These AMO in-
structions atomically load a data value from the address in rs1, place the value into register rd,
apply a binary operator to the loaded value and the original value in rs2, then store the result back
to the address in rs1
```

For atomic inc, If we want to return original content of the pointer and we know AMOADD succeed, than we can return `rd - rs2`.

**Review by AntonKozlov — CHANGES_REQUESTED** [review]
Was unable to look through whole PR, but there is an issue

File: src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp
Comment (DmKazakov) [reply src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp]: According to previous code, we want to return value after addition. Anyway, can we be sure that new value is still in `rd`? Can't someone change it again?

Comment (AntonKozlov) [reply src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp]: After we loaded a value from memory to register, only own core can change the value. Consider this like a private copy of a memory value

Comment (AntonKozlov) [reply src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp]: Please don't be blocked by this part. I think we first need unblock others and then polish these atomics to perfect state

File: src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp
Comment (DmKazakov) [reply src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp]: Fixed

File: src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp
Comment (DmKazakov) [reply src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp]: But we don't want to change the register value. We want to change value this register points to.

File: src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp
Comment (vasyoid) [inline src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp]: Do we actually need this label here?

Comment (DmKazakov) [reply src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp]: I don't know why it is needed. Fixed.

File: src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp
Comment (AntonKozlov) [inline src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp:106]: @azul-research/riscv Good for now, but there is idiomatic ways to do this.

To print use `tty` object. Like `tty->print_cr("...")`. It can format like `printf`: 
```
tty->print("some %d message part", 1);
if (something)
  tty->print("some other message part");
tty->print_cr("a message that ends a line");
```

A fatal error with a message is `fatal("message")` or `assert`. It will print a reason, file and line-number, generate hs_err file.... Like `assert(false, "untested")` here, will save a release build from crash also.

A crash by a reason is `Unimplemented` (self-describing), `ShouldNotReachHere` (a situation that can never happen during a normal execution), like
```
// int value;
if (value == 1) {
} else if (value == 2) {
else {
  ShouldNotReachHere(); // all other values are impossible here due to the rest of code
}
```

File: src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp
Comment (AntonKozlov) [inline src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp]: Please fix this.

`is_successful` register is written, so it should be in `out` list (here it's in `in`). It should be also have `"=&r"` constraint: it's written out (`=`) and it should differs from `in` registers (`&`). 

Without `&`, gcc is allowed to allocate `is_successful` to be equal to `dest`, for example. So the code will be equal to 
```
1:
lr.w    %[old_value], (%t0) // correct on 1st, incorrect on 2nd+ iteration
bne     %[compare_value], %[old_value], 2f
sc.w    %t0, %[exchange_value], (%t0) // destroys address to load from/store to
bnez    %t0, 1b
```

Such issue will be hard to debug down to this function. So we better spend some time to reviewing/fixing than to  considerably longer debugging later.

**Review by AntonKozlov — CHANGES_REQUESTED** [review]

File: src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp
Comment (DmKazakov) [reply src/hotspot/os_cpu/linux_riscv/atomic_linux_riscv.hpp]: Fixed

**Review by AntonKozlov — APPROVED** [review]

**Review by vasyoid — APPROVED** [review]

**Review by Geckostya — APPROVED** [review]

**Pull request merged**
